### PR TITLE
Feature/1205123214195391 inflation disinflation

### PR DESCRIPTION
**Pull request by talhadaar**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205123214195391

### CODE REVIEW

File: node/src/parachain/dev_chain_spec.rs [hunk 1]
```diff
@@ -140,6 +140,17 @@ fn configure_genesis(
 		staking_coefficient_reward_calculator: StakingCoefficientRewardCalculatorConfig {
 			coefficient: staking::coefficient(),
 		},
+		inflation_manager: InflationManagerConfig {
+			inflation_configuration: inflation_manager::InflationConfigurationT {
+				base_inflation_parameters: inflation_manager::InflationParametersT {
+					effective_inflation_rate: Perbill::from_percent(7 / 5), // 3.5%
```
Comment (sfffaaa) [inline node/src/parachain/dev_chain_spec.rs]: Sorry, I'm a little confused about the comment: what does the 3.5% mean, is that related to the Perbill::from_percent(7 / 5)?

File: pallets/inflation-manager/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+			let mut logic_switch: bool = false;
+
+			let current_year = CurrentYear::<T>::get();
+			let inflation_config = InflationConfiguration::<T>::get();
+			let mut inflation_parameters = YearlyInflationParameters::<T>::get();
+			let mut block_rewards = BlockRewards::<T>::get();
+
+			// if we have reached the stagnation year, kill the recalculation flag
+			// and set inflation parameters to stagnation values
+			if current_year == inflation_config.inflation_stagnation_year {
```
Comment (sfffaaa) [inline pallets/inflation-manager/src/lib.rs]: Does that mean once we reach the condition, `current_year == inflation_config.inflation_stagnation_year`, we'll do the logic_switch behavior below every block?

File: pallets/inflation-manager/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+			});
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> Pallet<T> {
+		// calculate inflationary tokens per block
+		fn rewards_per_block(inflation_parameters: &InflationParametersT) -> Balance {
```
Comment (sfffaaa) [inline pallets/inflation-manager/src/lib.rs]: In my understanding, we'll calculate the rewards_per_block once a year and store it in the blockchain. The block reward pallet will get the information from the blockchain for token minting, right?

**Comment by sfffaaa** [comment]
BTW, I saw some TODOs there; please remember to check that. And please also add some test cases for your inflation manager pallet.

File: pallets/inflation-manager/src/lib.rs (hunk 2 above)
Comment (talhadaar) [reply pallets/inflation-manager/src/lib.rs]: the logic_switch is a flag for wether BlockRewards per block need to updated, which happens only after each time YearlyInflationParameters are updated. So logic_switch behaviour happens only at end of year block.

File: node/src/parachain/dev_chain_spec.rs (hunk 1 above)
Comment (talhadaar) [reply node/src/parachain/dev_chain_spec.rs]: Thanks for that catch, it's supposed to be 7/2.

File: pallets/inflation-manager/src/lib.rs (hunk 3 above)
Comment (talhadaar) [reply pallets/inflation-manager/src/lib.rs]: BlockRewards are updated at end of the year, and the value is stored as a StorageValue in inflation-manager pallet, block reward pallet then fetches the value from there for minting.

File: pallets/inflation-manager/src/lib.rs (hunk 2 above)
Comment (talhadaar) [reply pallets/inflation-manager/src/lib.rs]: logic_switch was removed in favour of a more robust implementation.

**Review by sfffaaa — APPROVED** [review]
